GM: Escalade 2017 support

rickygilleland opened the pull request:
2017 Escalade port. Also adds support for fingerprinting v2 to GM vehicles.

- [x] added entry to CarInfo in selfdrive/car/*/values.py and ran `selfdrive/car/docs.py` to generate new docs
- [x] test route added to [routes.py](https://github.com/commaai/openpilot/blob/master/selfdrive/car/tests/routes.py)
- [x] route with openpilot: ef8f2185104d862e|2023-02-09--18-37-13
- [x] route with stock system: ef8f2185104d862e|2023-02-09--18-56-47

rickygilleland committed:
Initial 2017 Escalade values

rickygilleland committed:
update docs/CARS.md

sshane commented:
Seems very similar to the other Escalade we support, can you add your CAN fingerprint to it? We are planning on adding FPv2 support eventually, but probably not in a car port PR.

sshane commented on selfdrive/car/gm/values.py:
```diff
@@ -135,6 +138,28 @@ class CanBus:
+FW_QUERY_CONFIG = FwQueryConfig(
+  requests=[
+    Request(
+      [StdQueries.SHORT_TESTER_PRESENT_REQUEST, StdQueries.OBD_VERSION_REQUEST],
+      [StdQueries.SHORT_TESTER_PRESENT_RESPONSE, StdQueries.OBD_VERSION_RESPONSE],
```
I have some queries that the dealer uses to get the software part numbers, I just haven't figured out which to use, since one ECU has 10 or so different software codes. Thanks for attempting, but let's remove the FW code for now.

rickygilleland replied:
Removed and I'll hit you up on Discord.

rickygilleland committed:
revert back to v1 fingerprint & add test route

rickygilleland committed:
Merge branch 'master' into escalade

rickygilleland commented:
> Seems very similar to the other Escalade we support, can you add your CAN fingerprint to it and updating the car info to include your new car? We are planning on adding FPv2 support eventually, but probably not in a car port PR.

Yeah I had a feeling including it in a car port would be a little much. I reverted back to the CAN fingerprint and removed the other changes. Should I do a new test route using the v1 fingerprint, or are the test routes I already completed with the v2 fingerprint fine?

rickygilleland committed:
revert bad merge

sshane commented:
Firstly, do you think that we should make this a new platform, or is this similar enough with the ESV? We can just add your fingerprint to that list, and maybe tune the params a bit.

rickygilleland commented:
> Firstly, do you think that we should make this a new platform, or is this similar enough with the ESV? We can just add your fingerprint to the ESV's list, and maybe tune the params a bit.

Ah, I see what you're saying. I think they're different enough, but I'll defer to you if you feel otherwise. The ESV has a longer wheelbase and weighs more. When I was first testing out the port I was using the mass and wheelbase values from the ESV and the handling was really poor.

My assumption would be if we tune the ESV platform for better performance on my car, it'll affect the ESV performance negatively.

rickygilleland commented on selfdrive/car/gm/values.py:
```diff
@@ -174,6 +176,19 @@ class CanBus:
     190: 6, 193: 8, 197: 8, 199: 4, 201: 8, 208: 8, 209: 7, 211: 2, 241: 6, 249: 8, 288: 5, 289: 8, 298: 8, 304: 1, 309: 8, 313: 8, 320: 3, 322: 7, 328: 1, 338: 6, 340: 6, 352: 5, 381: 8, 384: 4, 386: 8, 388: 8, 393: 8, 398: 8, 413: 8, 417: 7, 419: 1, 422: 4, 426: 7, 431: 8, 442: 8, 451: 8, 452: 8, 453: 6, 454: 8, 455: 7, 462: 4, 463: 3, 479: 3, 481: 7, 485: 8, 489: 8, 497: 8, 499: 3, 500: 6, 501 ...
   }],
+  CAR.ESCALADE: [
+  {
+    193: 8, 197: 8, 201: 8, 208: 8, 209: 7, 211: 2, 241: 6, 249: 8, 298: 8, 304: 1, 309: 8, 311: 8, 313: 8, 320: 3, 322: 7, 328: 1, 352: 5, 381: 6, 386: 8, 388: 8, 413: 8, 451: 8, 452: 8, 453: 6, 455: 7, 460: 5, 463: 3, 479: 3, 481: 7, 485: 8, 489: 8, 497: 8, 500: 6, 501: 8, 510: 8, 532: 6, 534: 2, 560: 8, 562: 8, 563: 5, 608: 8, 609: 6, 610: 6, 611: 6, 612: 8, 613: 8, 707: 8, 761: 7, 801: 8, 810 ...
+  },
+  {
+    170: 8, 190: 6, 192: 5, 193: 8, 197: 8, 199: 4, 201: 8, 208: 8, 209: 7, 211: 2, 241: 6, 249: 8, 288: 5, 289: 1, 290: 1, 298: 8, 300: 1, 304: 8, 309: 8, 311: 8, 313: 8, 320: 8, 322: 7, 328: 1, 352: 7, 368: 8, 369: 8, 381: 5, 386: 5, 388: 8, 393: 7, 398: 8, 407: 4, 413: 8, 417: 7, 419: 1, 422: 4, 426: 7, 431: 8, 442: 8, 451: 8, 452: 8, 453: 6, 454: 8, 455: 7, 460: 5, 462: 4, 463: 3, 479: 3, 481 ...
+  },
+  {
+    170: 8, 190: 6, 192: 5, 193: 8, 197: 8, 199: 4, 201: 6, 208: 8, 209: 7, 211: 2, 241: 6, 249: 8, 288: 5, 289: 1, 290: 1, 298: 8, 300: 1, 304: 8, 309: 8, 311: 8, 313: 8, 320: 8, 322: 7, 328: 1, 352: 7, 368: 8, 369: 8, 381: 6, 386: 5, 388: 8, 393: 7, 398: 8, 407: 4, 413: 8, 417: 7, 419: 1, 422: 4, 426: 7, 431: 8, 442: 8, 451: 8, 452: 8, 453: 6, 454: 8, 455: 7, 460: 5, 462: 4, 463: 3, 479: 3, 481 ...
+  },
+  {
+    170: 8, 190: 6, 192: 5, 193: 8, 197: 8, 199: 4, 201: 8, 208: 8, 209: 7, 211: 2, 241: 6, 249: 8, 288: 5, 289: 1, 290: 1, 298: 8, 300: 1, 304: 8, 309: 8, 311: 8, 313: 8, 320: 8, 322: 7, 328: 1, 352: 7, 368: 8, 369: 8, 381: 5, 386: 5, 388: 8, 393: 7, 398: 8, 407: 4, 413: 8, 417: 7, 419: 1, 422: 4, 426: 7, 431: 8, 442: 8, 451: 8, 452: 8, 453: 6, 454: 8, 455: 7, 460: 5, 462: 4, 463: 3, 479: 3, 481 ...
+  }],
```
I'm also unsure if I should have this many fingerprints. I did try merging all of them together into a single dict, but each one had one or more overlapping keys with different values.

Verylukyguy replied:
There was also an issue fingerprinting the 2016 Escalade ESV and when the port was merged, it caused other GM vehicles to fingerprint as that vehicle.

sshane replied:
Are these all from different runs? You should only need to run the CAN fingerprinting script once, and copy the full dictionary.

As a reminder, it's best to start the script before the car even turns on, then start the car and let it run for a few minutes. Then turn the car off, back on, put it in drive, park, etc. The dict at that point should be pretty complete

sshane replied:
Did you get a new FP?

rickygilleland replied:
Those were all from a single run, but it's also possible I didn't let the car run long enough. 

I'll do another run tomorrow and let it go for 5-10 minutes and hopefully get a more complete fingerprint.

sshane replied:
These are multiple dicts, you should only get one from the script. How did you run it?

rickygilleland replied:
I ran it via the fingerprinting 1.0 instructions [here](https://github.com/commaai/openpilot/wiki/Fingerprinting#fingerprinting-10).

When I first ran it a few weeks ago, I copied only the last message, but that resulted in a car unrecognized error. When I ran it a second time, I noticed some of the keys were getting different values every few seconds and unless I included the multiple variants, it gave me the car unrecognized error. 

Same thing when I ran it a few days ago. Let it run for 2-4 minutes continuously, then turned it on/off a few times, switched gears, and let it run another minute. I then went back into the output and pulled the fingerprints that are currently in this PR. 

With all of those fingerprints present, it recognizes the car just fine. If I remove any of them (I tried), it fails to recognize the car.

Also fwiw, these fingerprints were all generated with the car stock per the instructions, but the same behavior seems to happen when I have my ASCM bypassed.

sshane replied:
I think you're just a little confused how the fingerprinting works. The can fingerprinting script reads the can traffic on the bus and records every message address and length it sees. Then during fingerprinting time, it will eliminate cars based on the addresses present on startup. If the car has an address not in a database fingerprint, it will eliminated that car.

So basically, you just need to just run the script with the car off, then turn it on and wait a few minutes. Any address it misses means it may fail to fingerprint later, so leaving it running for longer is best.

Then you only should be left with one dict once you're done (lengths shouldn't change, but you may have more or less addresses based on how long you let it run).

It's possible it's failing to fingerprint because your CAN fingerprint is identical to the Escalade. You can send me a route where it fails (with just one dict) and I can check.

koch-cf replied:
My experience with the fingerprinting so far was that it wouldn’t work with all 150 addresses that I gathered.  I saw that one of them did not appear until a few seconds after the car started, maybe because it has to boot up or something. So my theory is that OP was checking the FP before the last one popped onto the network. So I’ve been using the 149 and it works fine even if the car has already started and then OP boots. If I add the 150th one in it always fails. Hopefully my car doesn’t share those 149 with anything else. So anyway I would watch the log when starting the car and see which address may be popping up late, remove that from the fingerprint, and see if it works. Keep in mind that I’m a complete noob here so take my advice at your own risk :-)

sshane replied:
You want the opposite. You want to include as many addresses as possible, so that when you fingerprint that address doesn't eliminate your car.

sshane replied:
I updated the fingerprint from a route. Can you try to drive on the branch and make sure it fingerprints properly?

sshane replied:
Ah I see the issue. One address is on bus 0 and 2, but the length is different! Let me know when you get a drive on the latest commits I made

rickygilleland replied:
Sorry for the delay and thanks for all your help! Just pulled down your latest commit and started the car, but it didn't recognize it. Here is the route if that's helpful: `ef8f2185104d862e|2023-02-14--15-27-29--0`

sshane replied:
Okay try that, make sure to pull

rickygilleland replied:
Boom, that did it. Do you need updated routes?

Verylukyguy commented:
> Ah, I see what you're saying. I think they're different enough, but I'll defer to you if you feel otherwise. The ESV has a longer wheelbase and weighs more. When I was first testing out the port I was using the mass and wheelbase values from the ESV and the handling was really poor.
> 
> My assumption would be if we tune the ESV platform for better performance on my car, it'll affect the ESV performance negatively.

I agree.

The Chevy Tahoe, GMC Yukon, and Cadillac Escalade are shorter and lighter and the Chevy Suburban, Yukon XL, and Cadillac Escalade ESV are on the lengthened chassis and are heavier.  It might be possible to group them in this way, but the steering and suspensions might also be different.

rickygilleland committed:
Merge branch 'commaai:master' into escalade

sshane committed:
Merge remote-tracking branch 'upstream/master' into escalade

sshane committed:
update fingerprint

sshane committed:
update

sshane committed:
update

rickygilleland committed:
Merge branch 'commaai:master' into escalade

sshane commented on selfdrive/car/gm/values.py:
```diff
@@ -174,6 +176,10 @@ class CanBus:
   }],
+  CAR.ESCALADE: [
+  {
+    170: 8, 190: 6, 193: 8, 197: 8, 199: 4, 201: 8, 208: 8, 209: 7, 211: 2, 241: 6, 249: 8, 288: 5, 298: 8, 304: 1, 309: 8, 311: 8, 313: 8, 320: 3, 322: 7, 328: 1, 352: 5, 381: 6, 384: 4, 386: 8, 388: 8, 393: 7, 398: 8, 407: 4, 413: 8, 417: 7, 419: 1, 422: 4, 426: 7, 431: 8, 442: 8, 451: 8, 452: 8, 453: 6, 454: 8, 455: 7, 460: 5, 462: 4, 463: 3, 479: 3, 481: 7, 485: 8, 487: 8, 489: 8, 497: 8, 499 ...
```
```suggestion
    170: 8, 190: 6, 193: 8, 197: 8, 199: 4, 201: 8, 208: 8, 209: 7, 211: 2, 241: 6, 249: 8, 288: 5, 298: 8, 304: 1, 309: 8, 311: 8, 313: 8, 320: 3, 322: 7, 328: 1, 352: 5, 381: 6, 384: 4, 386: 8, 388: 8, 393: 7, 398: 8, 407: 4, 413: 8, 417: 7, 419: 1, 422: 4, 426: 7, 431: 8, 442: 8, 451: 8, 452: 8, 453: 6, 454: 8, 455: 7, 460: 5, 462: 4, 463: 3, 479: 3, 481: 7, 485: 8, 487: 8, 489: 8, 497: 8, 499: 3, 500: 6, 501: 8, 508: 8, 510: 8, 532: 6, 534: 2, 554: 3, 560: 8, 562: 8, 563: 5, 564: 5, 573: 1, 608: 8, 609: 6, 610: 6, 611: 6, 612: 8, 613: 8, 647: 6, 707: 8, 715: 8, 761: 7, 801: 8, 804: 3, 810: 8, 840: 5, 842: 5, 844: 8, 866: 4, 880: 6, 961: 8, 967: 4, 969: 8, 977: 8, 979: 8, 985: 5, 1001: 8, 1005: 6, 1009: 8, 1017: 8, 1019: 2, 1020: 8, 1033: 7, 1034: 7, 1105: 6, 1217: 8, 1221: 5, 1223: 2, 1225: 7, 1233: 8, 1249: 8, 1257: 6, 1265: 8, 1267: 1, 1280: 4, 1300: 8, 1322: 6, 1323: 4, 1328: 4, 1417: 8, 1609: 8, 1613: 8, 1649: 8, 1792: 8, 1798: 8, 1824: 8, 1825: 8, 1840: 8, 1842: 8, 1858: 8, 1860: 8, 1863: 8, 1872: 8, 1875: 8, 1882: 8, 1888: 8, 1889: 8, 1892: 8, 1906: 7, 1907: 7, 1912: 7, 1914: 7, 1917: 7, 1918: 7, 1919: 7, 1920: 7, 1937: 8, 1953: 8, 1968: 8, 2001: 8, 2017: 8, 2018: 8, 2020: 8, 2026: 8, 577: 8, 579: 8, 1345: 8, 1347: 8, 1512: 8, 1514: 8, 1601: 8, 1296: 4, 719: 5, 717: 5, 869: 4, 1930: 7
```

sshane committed:
Update selfdrive/car/gm/values.py

sshane commented on selfdrive/car/gm/values.py:
```diff
@@ -174,6 +176,10 @@ class CanBus:
   }],
+  CAR.ESCALADE: [
+  {
+    170: 8, 190: 6, 193: 8, 197: 8, 199: 4, 201: 8, 208: 8, 209: 7, 211: 2, 241: 6, 249: 8, 288: 5, 298: 8, 304: 1, 309: 8, 311: 8, 313: 8, 320: 3, 322: 7, 328: 1, 352: 5, 381: 6, 384: 4, 386: 8, 388: 8, 393: 7, 398: 8, 407: 4, 413: 8, 417: 7, 419: 1, 422: 4, 426: 7, 431: 8, 442: 8, 451: 8, 452: 8, 453: 6, 454: 8, 455: 7, 460: 5, 462: 4, 463: 3, 479: 3, 481: 7, 485: 8, 487: 8, 489: 8, 497: 8, 499 ...
```
```suggestion
    170: 8, 190: 6, 193: 8, 197: 8, 199: 4, 201: 8, 208: 8, 209: 7, 211: 2, 241: 6, 249: 8, 288: 5, 298: 8, 304: 1, 309: 8, 311: 8, 313: 8, 320: 3, 322: 7, 328: 1, 352: 5, 381: 6, 384: 4, 386: 8, 388: 8, 393: 7, 398: 8, 407: 4, 413: 8, 417: 7, 419: 1, 422: 4, 426: 7, 431: 8, 442: 8, 451: 8, 452: 8, 453: 6, 454: 8, 455: 7, 460: 5, 462: 4, 463: 3, 479: 3, 481: 7, 485: 8, 487: 8, 489: 8, 497: 8, 499: 3, 500: 6, 501: 8, 508: 8, 510: 8, 532: 6, 534: 2, 554: 3, 560: 8, 562: 8, 563: 5, 564: 5, 573: 1, 608: 8, 609: 6, 610: 6, 611: 6, 612: 8, 613: 8, 647: 6, 707: 8, 715: 8, 717: 5, 719: 5, 761: 7, 801: 8, 804: 3, 810: 8, 840: 5, 842: 5, 844: 8, 866: 4, 869: 4, 880: 6, 961: 8, 967: 4, 969: 8, 977: 8, 979: 8, 985: 5, 1001: 8, 1005: 6, 1009: 8, 1017: 8, 1019: 2, 1020: 8, 1033: 7, 1034: 7, 1105: 6, 1217: 8, 1221: 5, 1223: 2, 1225: 7, 1233: 8, 1249: 8, 1257: 6, 1265: 8, 1267: 1, 1280: 4, 1296: 4, 1300: 8, 1322: 6, 1323: 4, 1328: 4, 1417: 8, 1609: 8, 1613: 8, 1649: 8, 1792: 8, 1798: 8, 1824: 8, 1825: 8, 1840: 8, 1842: 8, 1858: 8, 1860: 8, 1863: 8, 1872: 8, 1875: 8, 1882: 8, 1888: 8, 1889: 8, 1892: 8, 1906: 7, 1907: 7, 1912: 7, 1914: 7, 1917: 7, 1918: 7, 1919: 7, 1920: 7, 1930: 7, 1937: 8, 1953: 8, 1968: 8, 2001: 8, 2016: 8, 2017: 8, 2018: 8, 2020: 8, 2024: 8, 2026: 8
```

sshane committed:
Update selfdrive/car/gm/values.py

sshane commented on selfdrive/car/gm/values.py:
```diff
@@ -174,6 +176,10 @@ class CanBus:
   }],
+  CAR.ESCALADE: [
+  {
+    170: 8, 190: 6, 193: 8, 197: 8, 199: 4, 201: 8, 208: 8, 209: 7, 211: 2, 241: 6, 249: 8, 288: 5, 298: 8, 304: 1, 309: 8, 311: 8, 313: 8, 320: 3, 322: 7, 328: 1, 352: 5, 381: 6, 384: 4, 386: 8, 388: 8, 393: 7, 398: 8, 407: 4, 413: 8, 417: 7, 419: 1, 422: 4, 426: 7, 431: 8, 442: 8, 451: 8, 452: 8, 453: 6, 454: 8, 455: 7, 460: 5, 462: 4, 463: 3, 479: 3, 481: 7, 485: 8, 487: 8, 489: 8, 497: 8, 499 ...
```
```suggestion
    170: 8, 190: 6, 193: 8, 197: 8, 199: 4, 201: 8, 208: 8, 209: 7, 211: 2, 241: 6, 249: 8, 288: 5, 298: 8, 304: 1, 309: 8, 311: 8, 313: 8, 320: 3, 322: 7, 328: 1, 352: 5, 381: 6, 384: 4, 386: 8, 388: 8, 393: 7, 398: 8, 407: 4, 413: 8, 417: 7, 419: 1, 422: 4, 426: 7, 431: 8, 442: 8, 451: 8, 452: 8, 453: 6, 454: 8, 455: 7, 460: 5, 462: 4, 463: 3, 479: 3, 481: 7, 485: 8, 487: 8, 489: 8, 497: 8, 499: 3, 500: 6, 501: 8, 508: 8, 510: 8, 532: 6, 534: 2, 554: 3, 560: 8, 562: 8, 563: 5, 564: 5, 573: 1, 608: 8, 609: 6, 610: 6, 611: 6, 612: 8, 613: 8, 647: 6, 707: 8, 715: 8, 717: 5, 719: 5, 761: 7, 801: 8, 804: 3, 810: 8, 840: 5, 842: 5, 844: 8, 866: 4, 869: 4, 880: 6, 961: 8, 967: 4, 969: 8, 977: 8, 979: 8, 985: 5, 1001: 8, 1005: 6, 1009: 8, 1017: 8, 1019: 2, 1020: 8, 1033: 7, 1034: 7, 1105: 6, 1217: 8, 1221: 5, 1223: 2, 1225: 7, 1233: 8, 1249: 8, 1257: 6, 1265: 8, 1267: 1, 1280: 4, 1296: 4, 1300: 8, 1322: 6, 1323: 4, 1328: 4, 1417: 8, 1609: 8, 1613: 8, 1649: 8, 1792: 8, 1798: 8, 1824: 8, 1825: 8, 1840: 8, 1842: 8, 1858: 8, 1860: 8, 1863: 8, 1872: 8, 1875: 8, 1882: 8, 1888: 8, 1889: 8, 1892: 8, 1906: 7, 1907: 7, 1912: 7, 1914: 7, 1917: 7, 1918: 7, 1919: 7, 1920: 7, 1930: 7, 1937: 8, 1953: 8, 1968: 8, 2001: 8, 2017: 8, 2018: 8, 2020: 8, 2026: 8
```

sshane committed:
no engine

rickygilleland committed:
Merge branch 'commaai:master' into escalade

sshane committed:
Merge remote-tracking branch 'upstream/master' into escalade

sshane commented on selfdrive/car/gm/interface.py:
```diff
@@ -198,6 +198,16 @@ def _get_params(ret, candidate, fingerprint, car_fw, experimental_long):
+      ret.wheelbase = 2.95  # 116 inches in meters
+      ret.steerRatio = 17.3
+      ret.centerToFront = ret.wheelbase * 0.5
+      ret.longitudinalTuning.kpV = [3.5, 2.18]
```
does the global tune not work very well?

rickygilleland replied:
It didn't at first, but that may no longer be the case. I'll do a test run without it and report back.

sshane replied:
what was the problem with it? late braking/acceleration? let me know if you think it's good enough and we can merge!

rickygilleland replied:
Late braking was the main issue. Just did a test run without it though and acceleration/stopping both seemed reasonable. 

Should be good to merge as far as I'm concerned!

sshane replied:
You can do something like this for the Acadia: https://github.com/commaai/openpilot/blob/89934b35d04291384b7a287a41ce49905f321167/selfdrive/car/gm/interface.py#L187

Seems like all GMs without regen control have a high delay to start braking. ICE and EVs without regen like the newer Bolt EVs

sshane replied:
Your longitudinal looks very similar to @Verylukyguy's. Try out that tuning option! In the mean time, this PR looks good to go. Thanks!

rickygilleland replied:
Ah, thank you! I'll give that a try in a few days when I have the bandwidth to do a proper check with PlotJuggler vs a more subjective check. 

If there's a significant change I can always open another PR in the future. 

It seems solid as is though either way. I didn't have to take over braking over about a 25 minute drive, so that's a significant improvement over when I initially tried to tune.

sshane replied:
Curious, how is steering? The plots look rather good. No oscillation, but could be a bit lazy it looks like.

rickygilleland replied:
Overall good. We took a 9 hour drive today and needed to help it on curves while at highway speed, and there was some minor oscillation instances where the road was sloped at an angle, but overall centered well!

rickygilleland committed:
Try global tune

sshane committed:
pcm

sshane commented on selfdrive/car/gm/values.py:
```diff
@@ -99,6 +100,7 @@ def init_make(self, CP: car.CarParams):
   CAR.MALIBU: GMCarInfo("Chevrolet Malibu Premier 2017"),
   CAR.ACADIA: GMCarInfo("GMC Acadia 2018", video_link="https://www.youtube.com/watch?v=0ZN6DdsBUZo"),
   CAR.BUICK_REGAL: GMCarInfo("Buick Regal Essence 2018"),
+  CAR.ESCALADE: GMCarInfo("Cadillac Escalade 2017", "Adaptive Cruise Control (ACC) & LKAS"),
```
```suggestion
  CAR.ESCALADE: GMCarInfo("Cadillac Escalade 2017", "Driver Assist Package"),
```

https://cdn.dealereprocess.org/cdn/brochures/cadillac/2017-escalade.pdf

sshane committed:
Driver Assist Package has ACC, and includes LKAS from previous package

sshane committed:
update docs

sshane commented on selfdrive/car/gm/interface.py:
```diff
@@ -198,6 +198,14 @@ def _get_params(ret, candidate, fingerprint, car_fw, experimental_long):
 
+    elif candidate == CAR.ESCALADE:
+      ret.minEnableSpeed = -1.  # engage speed is decided by pcm
+      ret.mass = 5653. * CV.LB_TO_KG + STD_CARGO_KG  # (5552+5815)/2
```
```suggestion
      ret.mass = 5653. * CV.LB_TO_KG + STD_CARGO_KG  # (5552 + 5815) / 2
```

sshane committed:
Merge remote-tracking branch 'upstream/master' into escalade

sshane committed:
add to releases
